src/components/WebPage.tsx

```diff
@@ -142,17 +148,60 @@ export default withRouter(
             } else {
                 this.props.history.replace(`${newUrl}`)
             }
+
+            this.undoPush(state, newUrl)
+        }
+
+        undoPush(state: Partial<WebPageState>, search: string): void {
+            const prevSearch = this.history[this.historyPosition - 1]?.search
+            if (prevSearch === search) {
+                return
+            }
+            this.history.splice(this.historyPosition)
+            this.history.push({ search, insertIndex: state?.insertIndex || 0 })
+            this.historyPosition = this.history.length
+        }
+
+        restoreSave(save: Save): void {
+            this.props.history.replace(save.search)
+            const state = this.getStateFromUrl(save.search)
+            state.insertIndex = save.insertIndex
+            this.setState(state)
+            this.log()
         }
 
-        rerunBuildOrder(gamelogic: GameLogic, buildOrder: Array<IBuildOrderElement>) {
+        undo = () => {
+            if (this.historyPosition === 1) {
+                return
+            }
+            this.historyPosition--
+            const save = this.history[this.historyPosition - 1]
+            this.restoreSave(save)
+        }
+
+        redo(): void {
+            if (this.historyPosition === this.history.length) {
+                return
+            }
+            const save = this.history[this.historyPosition]
+            this.historyPosition++
+            this.restoreSave(save)
+        }
+
```

Comment:
So this only saves a browser history entry if a new race is selected (= build order is cleared)?